src/message_bus.zig

```diff
@@ -18,38 +22,152 @@ pub const Message = struct {
+    connections_used: usize = 0,
+    /// Number of currently active replica connections in the connections slice.
+    replica_connections: usize = 0,
 
-    /// Messages that are ready to send:
-    ready: FIFO(Envelope) = .{},
+    /// Initialize the MessageBus for the given server replica and configuration.
+    pub fn init(
+        self: *MessageBus,
+        allocator: *mem.Allocator,
+        io: *IO,
+        server: *Replica,
+        server_id: u16,
+        configuration: []std.net.Address,
+    ) !void {
+        const replicas = try allocator.alloc(ReplicaConnection, configuration.len);
+        errdefer allocator.free(replicas);
 
-    const Address = union(enum) {
-        replica: *Replica,
-    };
+        for (configuration) |address, i| {
+            if (i == server_id) {
+                replicas[i].send_queue = .{};
+            } else {
+                replicas[i] = .{ .connection = .{
+                    .message_bus = self,
+                    .address = address,
+                } };
+                replicas[i].connect();
+            }
+        }
 
-    const Envelope = struct {
-        address: Address,
-        message: *Message,
-        next: ?*Envelope,
-    };
+        const connections = try allocator.alloc(Connection, num_connections);
+        errdefer allocator.free(connections);
+        mem.set(Connection, connections, .{ .message_bus = self });
 
-    pub fn init(allocator: *Allocator, configuration: []ConfigurationAddress) !MessageBus {
-        var self = MessageBus{
+        self.* = .{
             .allocator = allocator,
-            .configuration = configuration,
+            .io = io,
+            .server = server,
+            .server_fd = try init_tcp(configuration[server_id]),
+            .replicas = replicas,
+            .connections = connections,
         };
-        return self;
+
+        self.maybe_accept();
     }
 
-    pub fn deinit(self: *MessageBus) void {}
+    fn init_tcp(address: std.net.Address) !os.socket_t {
+        const fd = try os.socket(address.any.family, os.SOCK_STREAM | os.SOCK_CLOEXEC, 0);
+        errdefer os.close(fd);
+        // TODO: configure RCVBUF, SNDBUF, KEEPALIVE, TIMEOUT, NODELAY
+        try os.setsockopt(server, os.SOL_SOCKET, os.SO_REUSEADDR, &mem.toBytes(@as(c_int, 1)));
+        // TODO: port hopping
+        try os.bind(server, &address.any, address.getOsSockLen());
+        try os.listen(server, tcp_backlog);
+    }
+
+    /// Must be called whenever a Connection in the connections slice is closed.
+    fn maybe_accept(self: *MessageBus) void {
+        if (self.connections_used == self.connections.len) {
+            if (self.replica_connections == self.replicas.len - 1) {
+                // If there are no unusued connections and all replicas are
+                // connected, do nothing.
+                return;
+            } else {
+                assert(self.replica_connections < self.replicas.len - 1);
```

Comment:
We could skip the else branch here to reduce nesting and save indentation, since we already have the early return on line 103 and since we also have the assert on 105 in case the return was ever accidentally left out.